type-inference-in-readr-and-arrow.rmd

```diff
@@ -0,0 +1,271 @@
+
+In arrow, if you don’t want to individually specify column types, you can set up an option to import character columns as dictionaries (the equivalent of factors), which are converted into factors.
+
+```{r}
+dict_type <- data.frame(
+  x = c("yes", "no", "yes", "no")
+)
+
+write.csv(dict_type, "dict_type.csv", row.names = FALSE)
+
+# character data
+readr::read_csv("dict_type.csv")
+```
+
+```{r}
+# factor data
+readr::read_csv("dict_type.csv", col_types = list(x = col_factor()))
+```
+
+```{r}
+# set up the option. there's an open ticket to make this code a bit nicer to read.
+auto_dict_option <- arrow::CsvConvertOptions$create(auto_dict_encode = TRUE)
+arrow::read_csv_arrow("dict_type.csv", convert_options = auto_dict_option)
+```
+
+## Using schemas for manual control of data types
+
+Although relying on the reader itself to guess your column types can work well, what if you want more precise control?
+
+In readr, you can use the `col_types` parameter to specify column types. You can use the same parameter in arrow to use R type specifications.
+
+```{r}
+given_types <- data.frame(x = c(1, 2, 3), y = c(4, 5, 6))
+
+write.csv(given_types, "given_types.csv", row.names = FALSE)
+
+readr::read_csv("given_types.csv", col_types = list(col_integer(), col_double()))
+```
+
+You can also use this shortcode specification.
+
+```{r}
+readr::read_csv("given_types.csv", col_types = "id")
+```
+
+In arrow you can use the shortcodes (though not the `col_*()` functions), but you must specify the column names.
+
+We skip the first row as our data has a header row - this is the same behaviour as when we use both names and types in `readr::read_csv()` which then assumes that the header row is data if we don't skip it.
+
+```{r}
+arrow::read_csv_arrow("given_types.csv", col_names = c("x", "y"), col_types = "id", skip = 1)
+```
+
+What if you want to use Arrow types instead of R types though? In this case, you need to use a schema.  I won’t go into detail here, but in short, schemas are lists of fields, all of which contain a field name and a data type.  You can specify a schema like this:
+
+```{r}
+# this gives the same result as before - because our Arrow data has been converted to the relevant R type
+arrow::read_csv_arrow("given_types.csv", schema = schema(x = int8(), y = float32()), skip = 1)
+```
+
+```{r}
+# BUT, if you don't read it in as a data frame you'll see the Arrow type
+arrow::read_csv_arrow("given_types.csv", schema = schema(x = int8(), y = float32()), skip = 1, as_data_frame = FALSE)
+```
+
+## Parquet
+
+An alternative approach though, where you can specify your schema along with your data instead of when reading the data in, is to use Parquet format.  This means that if you’re sharing your data with others, you don’t need to worry about it being read in as the wrong data types.  In a follow-up post I’ll explore the Parquet format and compare management of data types in CSVs and Parquet.
```

Comment:
This first sentence is a little confusing:

```suggestion
An alternative approach is to use Parquet format, which stores the data types along with the data.  This means that if you’re sharing your data with others, you don’t need to worry about it being read in as the wrong data types.  In a follow-up post I’ll explore the Parquet format and compare management of data types in CSVs and Parquet.
```

Reply:
Much better, cheeers!